### PR TITLE
Add support for custom endpoints

### PR DESCRIPTION
**Pull request by alexander-demichev**
This PR adds support for custom endpoints, in order to do so we need to:
1. Fetch infrastructure object
2. Check if it contains custom endpoints in platform spec
3. Create custom endpoint resolver and use with AWS client 

https://docs.aws.amazon.com/sdk-for-go/v2/api/aws/endpoints/
https://github.com/openshift/enhancements/pull/163/files
https://github.com/openshift/installer/pull/3277/files

### CODE REVIEW

**Comment by alexander-demichev** [comment]
/hold 
I haven't tested the PR yet

File: pkg/client/client.go [hunk 1]
```diff
@@ -205,3 +214,44 @@ var addProviderVersionToUserAgent = request.NamedHandler{
...
+	if err := ctrlRuntimeClient.Get(context.Background(), infraName, infra); err != nil {
+		return err
+	}
+
+	// Do nothing when custom endpoints are missing
+	if infra.Spec.PlatformSpec.AWS == nil {
+		return nil
+	}
+
+	customEndpointsMap := buildCustomEndpointsMap(infra.Spec.PlatformSpec.AWS.ServiceEndpoints)
```
Comment (abhinavdahiya) [inline pkg/client/client.go:231]: Use the status and not the spec for these values.

Comment (enxebre) [reply pkg/client/client.go:231]: cc @alexander-demichev can we please address this?

Comment (alexander-demichev) [reply pkg/client/client.go]: fixed

File: pkg/client/client.go [hunk 2]
```diff
@@ -44,6 +46,8 @@ const (
 	AwsCredsSecretIDKey = "aws_access_key_id"
 	// AwsCredsSecretAccessKey is secret key containing AWS Secret Key
 	AwsCredsSecretAccessKey = "aws_secret_access_key"
+
+	globalInfrastuctureName = "cluster"
```
Comment (enxebre) [inline pkg/client/client.go:50]: can you reference where this magic "cluster" is coming from so others have context? i.e the line of code in the installer where this is instantiated

File: pkg/client/client.go [hunk 3]
```diff
@@ -205,3 +214,44 @@ var addProviderVersionToUserAgent = request.NamedHandler{
...
+	if err := ctrlRuntimeClient.Get(context.Background(), infraName, infra); err != nil {
+		return err
+	}
+
+	// Do nothing when custom endpoints are missing
+	if infra.Spec.PlatformSpec.AWS == nil {
+		return nil
+	}
+
+	customEndpointsMap := buildCustomEndpointsMap(infra.Status.PlatformStatus.AWS.ServiceEndpoints)
```
Comment (enxebre) [inline pkg/client/client.go:231]: this would panic if either `PlatformStatus` or `PlatformStatus.AWS` are nil?

Comment (alexander-demichev) [reply pkg/client/client.go:231]: we have a check on lines 227-229

File: pkg/client/client.go (hunk 2 above)
Comment (alexander-demichev) [reply pkg/client/client.go:51]: 👍

File: pkg/client/client.go (hunk 3 above)
Comment (enxebre) [reply pkg/client/client.go:232]: as far as I can see that's checking `infra.Spec.PlatformSpec.AWS`. That won't prevent this line from panicking.
I think the check above should include all the pointers that can be nil `infra.Status.PlatformStatus !=nil`, `infra.Status.PlatformStatus.AWS != nil`

Comment (alexander-demichev) [reply pkg/client/client.go:232]: `PlatformSpec` is not a pointer here, it can't be nil

Comment (enxebre) [reply pkg/client/client.go:232]: PlatformStatus is a pointer, PlatformStatus.AWS is another pointer.
You are dereferencing them without checking they are nil.
am I missing anything?

Comment (alexander-demichev) [reply pkg/client/client.go:232]: Sorry, you are right. I was looking at `PlatformSpec` :)  I pushed a fix for that.

File: pkg/client/client.go (hunk 2 above)
Comment (abhinavdahiya) [reply pkg/client/client.go:51]: https://github.com/openshift/api/blob/master/config/v1/types_infrastructure.go#L10

File: pkg/client/client.go [hunk 4]
```diff
@@ -205,3 +215,44 @@ var addProviderVersionToUserAgent = request.NamedHandler{
...
+	infraName := client.ObjectKey{Name: globalInfrastuctureName}
+
+	if err := ctrlRuntimeClient.Get(context.Background(), infraName, infra); err != nil {
+		return err
+	}
+
+	// Do nothing when custom endpoints are missing
+	if infra.Status.PlatformStatus == nil && infra.Status.PlatformStatus.AWS == nil {
+		return nil
+	}
```
Comment (abhinavdahiya) [inline pkg/client/client.go:230]: should this be OR ?

File: pkg/client/client.go [hunk 5]
```diff
@@ -205,3 +215,44 @@ var addProviderVersionToUserAgent = request.NamedHandler{
...
+		}
+		return endpoints.DefaultResolver().EndpointFor(service, region, optFns...)
+	}
+
+	awsConfig.EndpointResolver = endpoints.ResolverFunc(customResolver)
+
+	return nil
+}
+
+func buildCustomEndpointsMap(customEndpoints []configv1.AWSServiceEndpoint) map[string]string {
```
Comment (abhinavdahiya) [inline pkg/client/client.go:250]: nit: add comment that explains the key values of the returned map.

File: pkg/client/client.go [hunk 6]
```diff
@@ -205,3 +215,44 @@ var addProviderVersionToUserAgent = request.NamedHandler{
...
+		return err
+	}
+
+	// Do nothing when custom endpoints are missing
+	if infra.Status.PlatformStatus == nil && infra.Status.PlatformStatus.AWS == nil {
+		return nil
+	}
+
+	customEndpointsMap := buildCustomEndpointsMap(infra.Status.PlatformStatus.AWS.ServiceEndpoints)
+
```
Comment (abhinavdahiya) [inline pkg/client/client.go:233]: nit: you can probably just return nil here when the len of map is zero as there is not custom resolution required.

**Comment by enxebre** [comment]
PTAL @abhinavdahiya  for the final lgtm

**Pull request merged**